### PR TITLE
Change linear whitespace handling:

### PR DESCRIPTION
**Pull request by nbougalis**
When splitting a string, the existing code would compress linear whitespace away completely; this seems to not be intention of [RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2), which allows replacing any amount of linear whitespace with a single space.

Requesting an additional signoff from @vinniefalco.

### CODE REVIEW

**Comment by vinniefalco** [comment]
Tests?

**Comment by nbougalis** [comment]
Sheesh! Everything fails. That’ll teach me to do a PR on a Friday night!

**Comment by vinniefalco** [comment]
The tests would be nicer if you could tell by looking at them what the results should be for example
```
test( "apple,    pear,     peach", { "apple", "pear", "peach" });
```

**Comment by nbougalis** [comment]
> The tests would be nicer if you could tell by looking at them what the results should be for example
> 
> ```
> test( "apple,    pear,     peach", { "apple", "pear", "peach" });
> ```

Probably, however to exercise all possible combinations (no spaces, single space, multiple spaces, in both prefix & suffix form) would require a lot of entries. I believe that this automatically generated string is better from that standpoint, but it is more difficult to troubleshoot in the case of failure.

If you feel strongly enough about manual checks, I can make that change.

**Comment by vinniefalco** [comment]
I don't have strong feelings about it but to be honest I have not checked the HTTP spec to see if the change is correct

**Comment by vinniefalco** [comment]
The latest RFC is https://tools.ietf.org/html/rfc7231 btw

**Comment by vinniefalco** [comment]
\sigh you made me scrutinize the test. I think that the test is not exercising the modified code path. 

Consider this string:
    "the    rain, in    spain"

For the purpose of splitting into tokens we expect two entries:
    "the rain" and "in spain"

As you can see, the multiple whitespace in between words for each item in the list has been folded down to one whitespace. The test does not look like it exercises this case. The test also does not check that multiple whitespace inside a quoted string is preserved, i.e. the string `"the   rain", in    spain` should turn into `"the   rain"` and `in spain`.

**Comment by nbougalis** [comment]
Fair enough. I’ll migrate to explicit tests. 🖖

File: src/test/beast/beast_rfc2616_test.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        beast::xor_shift_engine rng(0x243F6A8885A308D3);
+
+        std::vector<std::string> words;
+
+        for (int i = 0; i != 64; ++i)
+            words.push_back("X-" + std::to_string(ripple::rand_int(rng, 100, 1000)));
+
+        std::string question;
+
+        for (auto w : words)
```
Comment (miguelportilla) [inline src/test/beast/beast_rfc2616_test.cpp]: `auto const& w`

File: src/ripple/beast/rfc2616.h [hunk 2]
```diff
@@ -257,7 +257,10 @@ split(FwdIt first, FwdIt last, Char delim)
         }
         else if (is_lws (*iter))
         {
-            ++iter;
+            e.append (1, ' ');
+            do
+                iter++;
+            while (iter != last && is_lws (*iter));
```
Comment (miguelportilla) [inline src/ripple/beast/rfc2616.h]: Alternate suggestion:

```
e.append (1, ' ');
while (++iter != last && is_lws (*iter));
```

File: src/ripple/beast/rfc2616.h [hunk 3]
```diff
@@ -257,7 +257,10 @@ split(FwdIt first, FwdIt last, Char delim)
         }
         else if (is_lws (*iter))
         {
-            ++iter;
+            e.append (1, ' ');
```
Comment (seelabs) [inline src/ripple/beast/rfc2616.h]: Instead of unconditionally adding a space, what do you think about:
```
            if (!e.empty())
                e.append (1, ' ');
```
That way we would only add a space when we are not starting a new token and we could save the `trim_left` call (i.e. replace `trim` with `trim_right`).

Comment (nbougalis) [reply src/ripple/beast/rfc2616.h]: I'm confused as to how that would work. Remember, we aren't splitting the word only at spaces (e.g. look for `split_comma` below, which splits on commas.

Comment (vinniefalco) [reply src/ripple/beast/rfc2616.h]: I have a better idea, lets get those tests up first so we can visualize the cases that we're discussing!

**Comment by vinniefalco** [comment]
>  to exercise all possible combinations 

You can keep the procedural tests, and just add a few select manual cases of literal strings. There's no need to "exercise all possible combinations" just make sure you get the corner cases and as close to 100% coverage as possible.

File: src/ripple/beast/rfc2616.h (hunk 3 above)
Comment (seelabs) [reply src/ripple/beast/rfc2616.h]: @nbougalis As written, this code collapses all linear white space into a single space. When we add a token to the result, we trim the left and right white space. I'm proposing we change this so when starting a new token we collapse all linear white space into nothing, so we no longer need the left trim.

I.e. there's no reason to add a space to a new token just to trim it away later.

Comment (thejohnfreeman) [reply src/ripple/beast/rfc2616.h:260]: Here's a different [implementation](https://gist.github.com/thejohnfreeman/d1e298e0d860a1bf539e6b0178ae05d9) that's shorter, with fewer else-branches, more DRY, and more early exits, but with a `goto`.

File: src/test/beast/beast_rfc2616_test.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            if (!question.empty())
+                question += ",";
+
+            question.append(ripple::rand_int(rng, 0, 3), ' ');
+            question.append(w);
+            question.append(ripple::rand_int(rng, 0, 3), ' ');
+        }
+
+        BEAST_EXPECT(beast::rfc2616::split_commas(question) == words);
```
Comment (thejohnfreeman) [inline src/test/beast/beast_rfc2616_test.cpp:54]: As the author of [autocheck](https://github.com/thejohnfreeman/autocheck), I appreciate the property-based testing you're trying to do here, but I think we're going to want some example-based tests too. Makes it easier for humans to understand the edge cases, and they're deterministic. If we want to use property-based testing, we should lean on a library that offers deterministic generators.

**Comment by miguelportilla** [comment]
@nbougalis I've written a [branch ](https://github.com/miguelportilla/rippled/commits/rfc2616) with a more extensive unit test, feel free to cherry pick.

**Comment by seelabs** [comment]
@miguelportilla Did you mean to close this?

**Comment by miguelportilla** [comment]
@seelabs Oops! I did not intend to close the PR.

File: src/ripple/beast/rfc2616.h (hunk 3 above)
Comment (nbougalis) [reply src/ripple/beast/rfc2616.h:260]: I appreciate that implementation @thejohnfreeman but I'm slightly allergic to `goto`. It _may_ be OK here, and I think it does help make the implementation a little clearer, but I need to think whether I want to adopt that approach.

**Comment by thejohnfreeman** [comment]
I just pushed a branch [`rfc2616`](https://github.com/thejohnfreeman/rippled/commits/rfc2616) on my personal fork. It rebases this PR onto 1.6.0-b2 (020b28580816a295185aef8832a8fe07ecc40de2), then applies these changes:

- @miguelportilla's [tests](https://github.com/ripple/rippled/pull/3143#issuecomment-608724603)
- @miguelportilla's [fix for building RocksDB](https://github.com/ripple/rippled/pull/3356)
- my [alternative implementation](https://github.com/ripple/rippled/pull/3143#discussion_r403290429) of `rfc2616::split` (after fixing bugs)

It passes all the tests.

**Comment by nbougalis** [comment]
Thanks @thejohnfreeman. I looked at your branch and while I am heavily allergic to `goto it's certainly used to good effect here.

With that said, I'm not sure how I want to handle this PR. I need to verify that the parsing changes I am proposing here are actually consistent with RFC2616 semantics, instead of consistent with how *I* think things should be parsed.

With that in mind, I'll close this PR for now and reopen, if necessary, at a later time.
